File: 算法/全排列/全排列/全排列.py
# https://leetcode.cn/problems/permutations/submissions/

class Solution:    
    def search(self, nums, index, ans):
        if index >= len(nums):
            ans.append(copy.deepcopy(nums))
            return
        for i in range(index, len(nums)):
            nums[i], nums[index] = nums[index], nums[i]
            self.search(nums, index + 1, ans)
            nums[i], nums[index] = nums[index], nums[i]
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.search(nums, 0, ans)
        return ans

File: 算法/全排列/全排列/全排列.cpp
// https://leetcode.cn/problems/permutations/

class Solution {
public:
    vector<vector<int>> ans;
    void search(vector<int> &nums, int index)
    {
        if (index >= nums.size())
        {
            ans.emplace_back(nums);
            return;
        }
        for (int i=index; i<nums.size(); i++)
        {
            swap(nums[index], nums[i]);
            search(nums, index + 1);
            swap(nums[index], nums[i]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        search(nums, 0);
        return ans;
    }
};